File: src/main/java/org/fountainmc/api/entity/data/hanging/LeashKnotData.java
package org.fountainmc.api.entity.data.hanging;

import javax.annotation.Nullable;

import org.fountainmc.api.entity.data.EntityData;

public interface LeashKnotData extends HangingEntityData {

    @Nullable
    EntityData getLeashedEntity();

    @Override
    default void copyDataFrom(EntityData data) {
        HangingEntityData.super.copyDataFrom(data);
        if (data instanceof HangingEntityData) {
            // if we ever get anything to copy add it here
        }
    }
}

File: src/main/kotlin/org/fountainmc/api/entity/data/hanging/LeashKnotData.kt
package org.fountainmc.api.entity.data.hanging

import org.fountainmc.api.entity.data.EntityData

interface LeashKnotData : HangingEntityData {

    val leashedEntity: EntityData?

    /**
     * Copy all of the mutable properties from the given data into this object.
     *
     * Doesn't copy [leashedEntity] information,
     * since that may involve interaction with other entities.
     */
    override fun copyDataFrom(data: EntityData) {
        super.copyDataFrom(data)
    }
}
